internal/pkg/skuba/replica/replica.go

```diff
@@ -0,0 +1,409 @@
+// And updates affinity rule to requiredDuringSchedulingIgnoredDuringExecution when
+// cluster size is equal or greater to replicas.
+func (r *Helper) UpdateNodes() error {
+	if err := r.deploymentsHelper(); err != nil {
+		return err
+	}
+
+	for _, deployment := range r.Deployments.Items {
+		if err := r.deploymentHelper(deployment.Name); err != nil {
+			return err
+		}
+		r.ReplicaSize = int(*r.Deployment.Spec.Replicas)
+		if r.ReplicaSize == 0 {
+			r.ReplicaSize = r.MinSize
+		}
+
+		nodes := r.ClusterSize
+		updated := false
+		var err error
+		switch {
+		case nodes >= r.ReplicaSize:
+			updated, err = r.replaceAffinity(preferred, required)
+			if err != nil {
+				return err
+			}
+		case nodes >= r.MinSize:
+			updated, err = r.replaceAffinity(required, preferred)
+			if err != nil {
+				return err
+			}
+		}
+		if !updated && nodes <= r.ReplicaSize {
+			// update replicas to trigger pod re-distribution when affinity is in
+			// preferredDuringSchedulingIgnoredDuringExecution.
+			if _, err := r.updateDeploymentReplica(r.ClusterSize - 1); err != nil {
+				return err
+			}
+			if _, err := r.updateDeploymentReplica(r.ReplicaSize); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// UpdateBeforeNodeDrains patches affinity rules and wait for pod running before node remove.
+// Patch will apply to deployment with preferredDuringSchedulingIgnoredDuringExecution affinity
+// when cluster size equal or less then replicas.
+func (r *Helper) UpdateBeforeNodeDrains() error {
+	if err := r.deploymentsHelper(); err != nil {
+		return err
+	}
+
+	for _, deployment := range r.Deployments.Items {
+		if err := r.deploymentHelper(deployment.Name); err != nil {
+			return err
+		}
+		r.ReplicaSize = int(*r.Deployment.Spec.Replicas)
+		if r.ClusterSize > r.ReplicaSize {
+			continue
+		}
+
+		updated, err := r.replaceAffinity(required, preferred)
+		if err != nil {
+			return err
+		}
+		if !updated && r.ClusterSize <= r.ReplicaSize {
+			// update replicas to trigger pod re-distribution when affinity is in
+			// preferredDuringSchedulingIgnoredDuringExecution.
+			if _, err := r.updateDeploymentReplica(r.ClusterSize - 1); err != nil {
+				return err
+			}
+			if _, err := r.updateDeploymentReplica(r.ReplicaSize); err != nil {
+				return err
+			}
+			updated = true
+		}
+		if updated {
+			var err error
+			retry := retryCount
+			err = r.waitForUpdates(&retry, &err)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *Helper) waitForUpdates(retry *int, e *error) error {
+	switch {
+	case *retry == 0:
+		return errors.Wrap(*e, "retry exhausted")
+	default:
+		if *e = r.waitForDeploymentReplicas(); *e == nil {
+			return *e
+		}
+		*retry--
+
+		klog.Warningf("waiting for deployment be available: %d", *retry)
+
+		time.Sleep(retryInterval)
+
+		if err := r.removePendingPods(); err != nil {
+			return err
+		}
+
+		if err := r.waitForUpdates(retry, e); err != nil {
+			return err
+		}
+	}
+	return nil
```

Comment:
```
	switch {
	case retry == 0:
		return errors.Wrap(*e, "retry exhausted")
	default:
		if err := r.waitForDeploymentReplicas(); err == nil {
			return nil
		}

 		klog.Warningf("waiting for deployment be available: %d", *retry)

 		time.Sleep(retryInterval)

 		if err := r.removePendingPods(); err != nil {
			return err
		}

 		if err := r.waitForUpdates(retry); err != nil {
			return err
		}
	}
	return nil
```

Reply:
Need `retry--` for recursive calls.

Reply:
Yep got your point :) I am for this now.